Allow models extending Model to be instantiated without full ActiveAndroid stack being initialized.

amay opened the pull request:
To unit test my classes that extend `Model`, I need to be able to instantiate them outside of the context where `Application` is available to pass into `ActiveAndroid#initialize`. I added two null checks in the `Cache` class to allow the models to be instantiated as POJOs. I would welcome feedback on other ways to allow this functionality.

amay committed:
allow model to be instantiated if initialize has not been called

vonloxley commented:
I think, that public methods shouldn’t fail silently. What’s your reason to not test the ORM-part of your model?

amay commented:
Thanks for taking a look at this so quickly.

The use case for this is that there may be business logic in the model that can be tested using light-weight (ie. fast) unit tests. For example, you could have this model:

``` java
package com.example;

import com.activeandroid.Model;
import com.activeandroid.annotation.Column;

public class SimpleModel extends Model {
    @Column(name = "name")
    String name;

    public String uppercaseName() {
        return name.toUpperCase();
    }
}
```

If you want to test the uppercaseName, ideally you can write this simple test:

``` java
package com.example;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class SimpleModelTest {
    private SimpleModel model;

    @Before
    public void setUp() {
        model = new SimpleModel();
    }

    @Test
    public void testUppercaseName() {
        model.name = "Example";
        assertEquals("EXAMPLE", model.uppercaseName());
    }
}
```

The changes in this PR allow this to work. Notice that I don't have to worry about calling `ActiveAndroid#initialize`, which requires an android Activity, which can't be instantiated in a test context unless you use libraries like [robolectric](http://pivotal.github.com/robolectric/) (additional complexity) or the built-in `ApplicationTestCase` (really slow). Additionally, even once you get past that, there are side effects of `ActiveAndroid#initialize` (like interacting with the database) that are not ideal for unit testing.

I agree that the case where the user tries to interact with the model without having initialized ActiveAndroid should be surfaced more clearly. Ideally, the first time the user interacted with the model in a way that requires ActiveAndroid to be setup, a runtime exception would be thrown with a meaningful message.

vonloxley commented:
Thanks for the detailed explanation. Personally I don’t mind using `ApplicationTestCase`, but then, I don’t do TDD. 

I’d prefer to raise a RuntimeException, i. e. `FrameworkNotInitialized` as well. Would it be enough for you to catch such an exception while instantiating your model? If not, setting a flag and only raising it once should be acceptable. 

Normal use wouldn’t be affected with this new exception and IMHO the quality of the framework would increase.

amay committed:
Revert "allow model to be instantiated if initialize has not been called"

This reverts commit 4ad8e666423ebbf1eec68dddf76b3e809991dee6.

amay committed:
add failing test case for lazy init

amay committed:
throw ActiveAndroidNotInitialized if not inited

amay committed:
lazy init model

1. Pull references to mTableInfo into getter that is in charge of
initializing field on first access.
2. Add model to cache on save or load, instead of at creation time.

amay committed:
Merge branch 'lazy-initialize'

amay commented:
I added an `ActiveAndroidNotInitialized` error that is thrown by `Cache` if `ActiveAndroid#initialize` has not been called. I also updated `Model` to lazy initialize the `mTableInfo` field and add itself to the cache on `save` and `loadFromCursor` instead of in the constructor. The effect of these changes is that you can instantiate a subclass of `Model` now without having to do any initialization, but the first time you try to have any interaction with the db, the `ActiveAndroidNotInitialized` error will be thrown. This seems like a good middle ground to me.

On a side note: the `Cache#getEntity` method could be made a lot more efficient with the use of a `Map` for `sEntities` instead of a `Set`. The map would let you do an immediate look up of the model by id instead of having to do a linear search. Is there a concern that different record types could have the same ID?

zach-klippenstein commented:
The cache uses a HashSet - there's no more linear searching going on than would be with a HashMap.

amay commented:
I was referring to the for loop in the `Cache#getEntity`. If that was a map then you could retrieve the model by id. That assumes that the id is unique across the different model types, which I'm not sure about with sqlite.

vonloxley commented:
:+1: I like it.

vonloxley commented:
The lazy initialization was effectively killing the unit-tests.

Could you please check vonloxley/ActiveAndroid@cf1fbf62d97d23edf5dcc31ddce6b6212b7682bc?

amay commented:
I cloned your repo, checked out cf1fbf6, and all the tests are passing for me. Can you tell me more about what type of test errors you are seeing and how you are running your tests? I'm running in intellij.

vonloxley commented:
vonloxley/ActiveAndroid@cf1fbf6 fixes the tests. Sorry, if I didn’t make that clear. `OpenDatabase` calls methods in it’s own class, which in return call `checkInitialization`. Therefore `sIsInitialized` needs to be true before `openDatabase` is called…

amay commented:
Okay. That makes sense. Thanks for clarifying. If I swap those two lines I see the failures as well. The weird thing is that the tests don't break in my repo, so I didn't see those failures. Not sure why the tests would run any differently.

amay committed:
set initialized flag before opening database

vsigler commented:
The discussion here seems to have died, but just in case someone needs/wants to unit test stuff around ActiveAndroid without having to run on a live device/emulator or using Robolectric, here is my solution to the problem.

Regular mocking frameworks cannot mock away static stuff, but the PowerMock library can. This library has bytecode/classloader/classpath manipulating capabilities making this possible. Here is a base test class I use in my tests: https://github.com/vsigler/ActiveAndroid-Validation/blob/master/src/test/java/cz/sigler/android/aavalidation/TestBase.java

Now, in the actual tests I can instantiate models and also verify actual db calls and other things. Slight downside of this approach is that the tests are a bit slower.
